x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/TokensInvalidationResult.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.security.authc.support;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * The result of attempting to invalidate one or multiple tokens. The result contains information about:
+ * <ul>
+ * <li>how many of the tokens were actually invalidated</li>
+ * <li>how many tokens are not invalidated in this request because they were already invalidated</li>
+ * <li>how many tokens were not invalidated because of an error and what the error was</li>
+ * </ul>
+ */
+public class TokensInvalidationResult implements ToXContentObject {
+
+    private final String[] invalidatedTokens;
+    private final String[] prevInvalidatedTokens;
+    private final String[] errors;
+    private final int attemptCounter;
+
+
+    public TokensInvalidationResult(String[] invalidatedTokens, String[] notInvalidatedTokens,
+                                    @Nullable String[] errors, int attemptCounter) {
+        Objects.requireNonNull(invalidatedTokens, "invalidated_tokens must be provided");
+        this.invalidatedTokens = invalidatedTokens;
+        Objects.requireNonNull(notInvalidatedTokens, "not_invalidated_must be provided");
+        this.prevInvalidatedTokens = notInvalidatedTokens;
+        if (null != errors) {
+            this.errors = errors;
+        } else {
+            this.errors = new String[0];
+        }
+        this.attemptCounter = attemptCounter;
+    }
+
+    public static TokensInvalidationResult emptyResult(){
+        return new TokensInvalidationResult(new String[0], new String[0], new String[0], 0);
+    }
+
+    public String[] getInvalidatedTokens() {
+        return invalidatedTokens;
+    }
+
+    public String[] getPrevInvalidatedTokens() {
+        return prevInvalidatedTokens;
+    }
+
+    public String[] getErrors() {
```

Comment:
rather than strings, I wonder if we should have exceptions?

Reply:
Sure thing. If you have any suggestions on how the exceptions can be optimally presented to the user I'm all :ear:  ( I was thinking to only show `getMessage()` and this is why this was originally a `String[]`

Reply:
`ElasticsearchException#generateThrowableXContent` should do the trick

Reply:
TIL ! It'd still be tricky to implement equals on `TokensInvalidationResult` and `InvalidateTokenResult` when I can't compare the `ElasticsearchException`s byt I guess their string representations can be compared instead